src/linker/Linker.Steps/EmbeddedXmlStep.cs

```diff
@@ -86,8 +88,8 @@ protected override void Process ()
 									.Where (res => res.Name.Equals ("ILLink.Substitutions.xml", StringComparison.OrdinalIgnoreCase))
 									.Cast<EmbeddedResource> ()) {
 					try {
-						Context.LogMessage ($"Processing embedded substitution descriptor {rsc.Name} from {asm.Name}");
-						steps_to_add.Push (GetExternalSubstitutionStep (rsc, asm));
+						Context.LogMessage ($"Processing embedded substitution descriptor {rsc.Name} from {assembly.Name}");
+						steps.Add (GetExternalSubstitutionStep (rsc, assembly));
```

Comment:
Do we want to build in some mechanism to be able to run the substitutions step early? We will need this for the constant propagation work - there we need to resolve assembly and run substitutions, but avoid running descriptors (and attributes don't matter one way or another right now).

Reply:
I thought about this too - it might be nice for the required processing to be triggered directly when we ask questions that require it. For example`HasLinkerAttribute` could trigger scanning for embedded attribute XML and traversing the method body could ensure substitutions from that assembly have run. This way wouldn't require a call to `ProcessAssemblySteps` at key points in MarkStep.

The downside is that it would be a little more complicated if we want to avoid scanning resources multiple times (once each for attribute/substitutions/descriptor XML), and it would take separate caches for each assembly step that can be run separately - but it's not really that hard. I just went with this approach for now because it seemed simpler.

Could you clarify this?
> there we need to resolve assembly and run substitutions, but avoid running descriptors

I thought that the descriptors would need to be processed anyway - for example even if an assembly is only used via a call to a constant-returning method, that call would still be kept in the final output, so I'd expect the descriptors to be processed. If they need to be processed anyway, it reduces the benefit of letting these steps run separately.

Reply:
About not running descriptors:
I think the rule was, don't do anything on an assembly until it's actually marked. To correctly implement constant propagation we need to break that a little bit and say "It's OK to process substitutions on it". But having to load an assembly from the constant propagation doesn't mean it will be marked. Maybe it's in a branch which will be removed. I think it's gonna take a lot of changes to make the constant propagation perform branch removal in one pass (I looked into it, the code currently is in no way going to work even remotely like that, it would probably require complete rewrite and it's pretty tricky to do so).

I think it would also be a good idea to separate the concerns - constant propagation and marking. If we do process descriptors before the assembly is marked, the descriptor processing may end up marking more stuff than we need.

Reply:
How much do we actually rely on the fact that constant propagation can propagate things recursively?

I.e. that we can tell from `int Foo => IntPtr.Size; int Bar => Foo` that the `Bar` is constant. I understand that in theory it will help, but is it really helping enough to justify the engineering complexities?

Reply:
Not much - I just tried and the deepest is 3 - so basically the sample you have. For example `System.Environment.Is64BitOperatingSystem` is determined as constant, but to do that we need to recurse 2 more levels.

That said the problem discussed here is not about recursion really. Single level propagation is enough to trigger it, the problem is about cross-assembly. I don't have the numbers for that - but basically any `EventSource` in non-CoreLib assembly will run into it (`EventSource.IsEnabled` is substituted and is called from many other assemblies).

Some fun statistics (simple hello world, but we analyze big part of the framework under the current design):
* 3 iterations - so we run over all methods 3 times
* 72K methods analyzed (x 3) - this is the perf killer right now - we analyze 217K methods in total - LOT of them without any result and lot of them will be trimmed in the app
* 11 methods substituted (as in there 11 methods which were encountered and are substituted via the XML)
* 44 methods analyzed as constant after propagation (so at least one recursion)
* 1577 methods analyzed as constant in total
* 3183 callsites of constant methods found (probably exaggerated by 3 iterations, so it's more like 1K+ in reality)

Reply:
Thank you! Really eye-opening to see the statistics!

Reply:
I decoupled the steps from each other so that it's easy to run just substitutions or other steps (and to avoid unnecessary descriptor processing).
- Right now substitutions are triggered when we ask for a method action (see [GetMethodAction](https://github.com/sbomer/linker/blob/f61d38cb9f766dc4ffa2e27bbf69383633abde14/src/linker/Linker.Steps/MarkStep.cs#L2488))
- Constant propagation is requested directly from MarkMethodBody. When we know how to solve the constant prop problem we might be able to put this inside of a helper.